lib/url-decorator.jsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import { CompositeDecorator } from 'draft-js';
+import urlRegex from 'url-regex';
+import normalizeUrl from 'normalize-url';
+import { viewExternalUrl } from './utils/url-utils';
+
+const urlRe = urlRegex();
+/*
+- markdownLinkRe is a regexp to be applied to the substring preceding a matched url, ie. `wholeString.slice(0, wholeString.indexOf(matchedUrl))`
+- it tests whether the matched url is preceded by the charcters that define a link in markdown, ie. `(` or `]: `
+*/
+const markdownLinkRe = /(?:\(|\]:\s+)$/;
+
+export function normalizeUrlsInMarkdownString( str ) {
+	return str.replace( urlRe, ( match, offset ) => {
+		var normalized = normalizeUrl( match );
+		if ( markdownLinkRe.test( str.slice( 0, offset ) ) ) {
+			return normalized;
+		}
+		// replace url with markdown link only if it's not already part of one
+		return `[${match}](${normalized})`;
+	} );
+}
+
+export const urlDecorator = new CompositeDecorator( [
+	{
+		strategy: function( contentBlock, callback ) {
+			const text = contentBlock.getText();
+			let matchArr, start;
+			while ( ( matchArr = urlRe.exec( text ) ) !== null ) {
+				start = matchArr.index;
+				callback( start, start + matchArr[0].length );
+			}
+		},
+
+		component: React.createClass( {
+			onClick: function( event ) {
+				event.preventDefault();
+				viewExternalUrl( normalizeUrl( this.props.decoratedText ) );
+			},
+
+			render: function() {
+				return (
+					<a
+						onClick={this.onClick}
+						href={this.props.decoratedText}>
+						{this.props.children}
+					</a>
+				);
+			}
+		} ),
```

Comment:
It would be really helpful to have this component created elsewhere. Maybe just in the module scope of this file, but possible in another file altogether. Further, if we simply export the _strategy_ from here it's trivial enough to create the decorator in the note content editor itself, which would have certain advantages for code comparability and organization.